tools/testrunner/testrunner.go

```diff
@@ -0,0 +1,348 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package testrunner
+
+import (
+	"context"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"iter"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/jstemmer/go-junit-report/v2/junit"
+)
+
+type node struct {
+	children map[string]node
+}
+
+func (n node) visitor(path ...string) func(yield func(string, node) bool) {
+	return func(yield func(string, node) bool) {
+		for name, child := range n.children {
+			path := append(path, name)
+			if !yield(strings.Join(path, "/"), child) {
+				return
+			}
+			child.visitor(path...)(yield)
+		}
+	}
+}
+
+func (n node) walk() iter.Seq2[string, node] {
+	return n.visitor()
+}
+
+type attempt struct {
+	runner       *runner
+	number       int
+	junitXmlPath string
+	exitErr      *exec.ExitError
+	suites       junit.Testsuites
+}
+
+func (a *attempt) run(ctx context.Context, args []string) error {
+	args = append([]string{"--junitfile", a.junitXmlPath}, args...)
+	log.Printf("starting test attempt %d with args: %v", a.number, args)
+	cmd := exec.CommandContext(ctx, a.runner.gotestsumExecutable, args...)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	cmd.Stdin = os.Stdin
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *attempt) recordResult(exitErr *exec.ExitError) error {
+	var suites junit.Testsuites
+
+	f, err := os.Open(a.junitXmlPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	decoder := xml.NewDecoder(f)
+	if err := decoder.Decode(&suites); err != nil {
+		return err
+	}
+	a.exitErr = exitErr
+	a.suites = suites
+	return nil
+}
+
+func (a attempt) failures() []string {
+	root := node{children: make(map[string]node)}
+
+	for _, suite := range a.suites.Suites {
+		if suite.Failures == 0 {
+			continue
+		}
+		for _, tc := range suite.Testcases {
+			if tc.Failure == nil {
+				continue
+			}
+			n := root
+			for _, part := range strings.Split(tc.Name, "/") {
+				child, ok := n.children[part]
+				if !ok {
+					child = node{children: make(map[string]node)}
+					n.children[part] = child
+				}
+				n = child
+			}
+		}
+	}
+
+	leafFailures := make([]string, 0)
+
+	for path, n := range root.walk() {
+		if len(n.children) > 0 {
+			continue
+		}
+		leafFailures = append(leafFailures, path)
+	}
+
+	return leafFailures
+}
+
+type runner struct {
+	gotestsumExecutable string
+	junitOutputPath     string
+	attempts            []*attempt
+	retries             int
+}
+
+func newRunner(gotestsumExecutable string) *runner {
+	return &runner{
+		gotestsumExecutable: gotestsumExecutable,
+		attempts:            make([]*attempt, 0),
+	}
+}
+
+func (r *runner) sanitizeAndParseArgs(args []string) ([]string, error) {
```

Comment:
I'm surprised there is no arg parser in go stdlib :)

Reply:
There is, it's called `flag` but we need to sanitize and replace args as well, and that's not supported. We also cross the `--` delimiter so I ended up just writing this whole custom function.

Reply:
there is `flag.Parse()`. I don't know why Roey doesn't want to use it.